README.md
# D3 gauge

A simple gauge that uses d3.

You can have maximum and minimum values stored in this gauge.

The gauge can autoscale.

To see the gauge in action, download the repo, npm install, then run it locally with webpack to see what it looks like.

## Example gauge

![an example gauge image not here](gaugeexample.png)

## Example code

Here is an example code snippet:

```javascript
var gauge = new Gauge('#idOfSelectedElement', {
    size: 300,
    clipWidth: 300,
    clipHeight: 300,
    ringWidth: 60,
    maxValue: 10,
    transitionMs: 4000
});
gauge.render();
```

## Full range of config settings

```javascript
{
            size: 200,
            clipWidth: 200,
            clipHeight: 110,
            ringInset: 20,
            ringWidth: 20,

            pointerWidth: 10,
            pointerTailLength: 5,
            pointerHeadLengthPercent: 0.9,

            minValue: 0,
            maxValue: 10,

            minAngle: -90 - 45,
            maxAngle: 90 + 45,

            transitionMs: 750,

            majorTicks: 10,
            labelFormat: d3.format(',g'),
            labelInset: 10,
            // sets whether values when they overflow from maximum should be
            // snapped back to maximum or go back around from the start
            isCyclic: false,

            arcColorFn: d3.interpolateHsl(d3.rgb('#e8e2ca'), d3.rgb('#3e6c0a'))
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.3.17](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.15...v1.3.17) (2020-04-17)

Fixed up process from master.

### [1.3.16](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.15...v1.3.16) (2020-04-17)

### [1.3.15](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.14...v1.3.15) (2020-04-17)

Fixed more bugs and added update for some npm packages.

### [1.3.14](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.13...v1.3.14) (2020-04-17)

Added in security fixes for long overdue bugs.

### [1.3.13](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.12...v1.3.13) (2020-02-17)

Fixed security vulnerability in dependency.

### [1.3.12](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.11...v1.3.12) (2019-12-29)

Fixed security vulnerability in dependency - serialize-javascript

### [1.3.11](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.10...v1.3.11) (2019-12-09)

Fixed up vulnerabilities.

### [1.3.10](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.9...v1.3.10) (2019-12-01)

Added more jest tests to improve the coverage report.

### [1.3.9](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.8...v1.3.9) (2019-12-01)

Added in changes to the pointer to stop overflow values happening with the scale.

### [1.3.8](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.7...v1.3.8) (2019-11-27)

Added more content to the readme to explain how to use with a better example.

### [1.3.7](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.6...v1.3.7) (2019-11-26)

Small label code refactor in the labels to make the code more readable.

### [1.3.6](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.5...v1.3.6) (2019-11-26)

Fixed up the `npm package.json`

### [1.3.5](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.3...v1.3.5) (2019-11-26)

### [1.3.3](https://github.com/apollowebdesigns/d3gauge/compare/v1.3.1...v1.3.3) (2019-11-26)

### 1.3.1 (2019-11-26)

First changelog to be made for the project


src/gauge/gauge.js
const d3 = require('d3');
const Bar = require('./bar/bar');
const Pointer = require('./pointer/pointer');
const Labels = require('./labels/labels');

module.exports = class Gauge {
    constructor (container, configuration){
        this.config = {
            size						: 200,
            clipWidth					: 200,
            clipHeight					: 110,
            ringInset					: 20,
            ringWidth					: 20,

            pointerWidth				: 10,
            pointerTailLength			: 5,
            pointerHeadLengthPercent	: 0.9,

            minValue					: 0,
            maxValue					: 10,

            minAngle					: -90 - 45,
            maxAngle					: 90 + 45,

            transitionMs				: 750,

            majorTicks					: 10,
            labelFormat					: d3.format(',g'),
            labelInset					: 10,
            isCyclic                    : false,

            arcColorFn					: d3.interpolateHsl(d3.rgb('#e8e2ca'), d3.rgb('#3e6c0a'))
        };
        this.range = undefined;
        this.r = undefined;
        this.pointerHeadLength = undefined;
        this.value = 0;

        this.svg = undefined;
        this.arc = undefined;
        this.arcs = undefined;
        this.foregroundArc = undefined;
        this.backgroundArc = undefined;
        this.scale = undefined;
        this.ticks = undefined;
        this.tickData = undefined;
        this.pointer = undefined;

        this.pi = Math.PI;
        this.cur_color = 'limegreen';
        this.max = 180;
        this.min = 0;
        this.current = 10;
        this.current = 10;

        this.oR = 100;
        this.iR = 100 - this.config.ringWidth;

        this.donut = d3.scaleLinear();

        this.container = container;
        this.configure(configuration);
        this.redBar = (30 / 360) * 2 * Math.PI;
        this.orangeBar = (15 / 360) * 2 * Math.PI;
        this.changeTextColor = false;
        this.formatTicks = d3.format('.0f');
    }

    deg2rad(deg) {
        return deg * Math.PI / 180;
    }

    rad2deg(rad) {
        return rad * 180 / Math.PI;
    }

    newAngle(d) {
        let that = this;
        let ratio = that.scale(d);
        let newAng = that.config.minAngle + (ratio * that.range);
        return newAng;
    }

    configure(configuration) {
        let that = this;
        let prop = undefined;
        for ( prop in configuration ) {
            this.config[prop] = configuration[prop];
        }

        this.range = this.config.maxAngle - this.config.minAngle;
        this.r = this.config.size / 2;
        this.pointerHeadLength = Math.round(this.r * this.config.pointerHeadLengthPercent);

        // a linear scale that maps domain values to a percent from 0..1
        this.scale = d3.scaleLinear()
            .range([0,1])
            .domain([this.config.minValue, this.config.maxValue]);

        this.ticks = this.scale.ticks(this.config.majorTicks);
        this.tickData = d3.range(this.config.majorTicks).map(function() {return 1/that.config.majorTicks;});

        that.arc = d3.arc()
            .innerRadius(that.r - that.config.ringWidth - that.config.ringInset)
            .outerRadius(that.r - that.config.ringInset)
            .startAngle(function(d, i) {
                let ratio = d * i;
                return that.deg2rad(that.config.minAngle);
            })
            .endAngle(function(d, i) {
                let ratio = d * (i+1);
                return that.deg2rad(that.config.maxAngle);
            });
    }

    centerTranslation() {
        let that = this;
        return 'translate('+ that.r +','+ that.r +')';
    }

    isRendered() {
        let that = this;
        return (that.svg !== undefined);
    }

    addBackgroundArcToSvgParent(parentSvg) {
        return parentSvg.append('g')
            .attr('class', 'arc')
            .attr('transform', this.centerTranslation())
            .append('path')
            .datum({
                startAngle: -90 * (Math.PI / 180),
                endAngle: -90 * (Math.PI / 180)
            })
            .style('fill', 'red')
            .attr('d', this.arc);
    }

    render(newValue) {
        let that = this;
        that.svg = d3.select(that.container)
            .append('svg:svg')
            .attr('class', 'Gauge')
            .attr('width', that.config.clipWidth)
            .attr('height', that.config.clipHeight)
            .append('g');

        that.arcs = new Bar(that.svg, that.config, that.r, that.redBar, that.orangeBar, that.changeTextColor);
        that.labels = new Labels(that.svg, that.config, that.ticks);
        that.pointer = new Pointer(that.svg, that.config, that.r);
    }

    update(newValue) {
        this.pointer.update(newValue);
    }

    updateBar(newMax, newMin) {
        this.arcs.update(newMax, newMin);
    }

    updateLabels(newMax, newMin){
        this.labels.update(newMax, newMin);
    }
};
